src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -231,28 +252,67 @@ case class WorkflowActor(workflow: WorkflowDescriptor,
     }
 
     def lookup(identifierString: String): WdlValue = {
-      /* Start by assuming the identifierString is a Namespace reference, and if it is return that reference.
-       * Otherwise, assume it's referencing a Call and return the value of that call as a WdlObject,
-       * which represents the outputs of the Call (WdlObject only returned if outputs are present)
+      /* This algorithm defines three ways to lookup an identifier in order of their precedence:
+       *
+       *   1) Look for a WdlNamespace with matching name
+       *   2) Look for a Call with a matching name (perhaps using a scope resolution algorithm)
+       *   3) Look for a Declaration with a matching name (perhaps using a scope resolution algorithm)
+       *
+       * Each method is tried individually and the first to return a Success value takes precedence.
        */
 
-      val namespaces = workflow.namespace.namespaces filter {_.importedAs.contains(identifierString)}
-      namespaces.headOption.getOrElse {
-        val matchedCall = parentWorkflow.calls.find {_.name == identifierString}.getOrElse {
-          throw new WdlExpressionException(s"Expecting to find a call with name '$identifierString'")
+      def lookupNamespace(name: String): Try[WdlNamespace] = {
+        workflow.namespace.namespaces.find {_.importedAs.contains(name)} match {
+          case Some(x) => Success(x)
+          case _ => Failure(new WdlExpressionException(s"Could not resolve $identifierString as a namespace"))
         }
-        val futureValue = fetchCallOutputEntries(matchedCall).map { entries =>
-          val callOutputs = entries.map { entry =>
-            val value = entry.wdlValue match {
-              case Some(v) => v
-              case _ => throw new WdlExpressionException(s"Could not evaluate call '${matchedCall.name}', because '${entry.key.name}' is undefined")
+      }
+
+      def lookupCall(name: String): Try[WdlObject] = {
+        parentWorkflow.calls.find {_.name == identifierString} match {
+          case Some(matchedCall) =>
+            val futureValue = fetchCallOutputEntries(matchedCall).map {callOutputEntries =>
+              val callOutputsAsMap = callOutputEntries.map { entry =>
+                entry.key.name -> entry.wdlValue
+              }.toMap
+              callOutputsAsMap.find {case (k,v) => v.isEmpty} match {
+                case Some(noneValue) => Failure(new WdlExpressionException(s"Could not evaluate call ${matchedCall.name} because some if its inputs are not defined (i.e. $noneValue._1"))
+                case None => Success(WdlObject(callOutputsAsMap.map {case (k,v) => k -> v.get}))
+              }
             }
-            entry.key.name -> value
-          }
-          WdlObject(callOutputs.toMap)
+            Await.result(futureValue, DatabaseTimeout)
```

Comment:
It was blocking before... when we're in the middle of evaluating an expression I don't think we can be asynchronous...